### PR TITLE
Direct velocity primitive class

### LINKED ISSUE #91: Create DirectVelocity Primitive class
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
We need a class in the AI for the DirectVelocity Primitive, so the AI can use it as an output.

**Describe the solution you'd like**
Implement a DirectVelocity Primitive class that extends from the base Primitive class. This will go in the `thunderbots/software/ai/primtiive` folder. See the `move_primitive.*` files for an example.

Make sure to document the spec of the primitive, using the information from #86. This can probably go at the top of the class (in a big javadoc style comment).

### PR DESCRIPTION
**Pull request by HaomiaoZ**
### Description
Create Direct Velocity Primitive class as extension of Primitive class. Similar to move primitive, two types of constructor are created and all the functions in the primitive class are overridden. 

### Testing Done
Unit test can be complied through `catkin_make` , and all unit tests pass.

### Resolved Issues
resolves #91 

### Review Checklist
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1: 

Looping in @henrymliu and @EvanMorcom for a second look/sanity check.

**Comment by EvanMorcom** [comment]
I'm not an expert in the low-level but looks good to me. Only thing I see is the comments at the start of the file are missing.

**Review by henryymliu — CHANGES_REQUESTED** [review]
Looks good, some minor documentation-related comments.

File: src/thunderbots/software/ai/primitive/direct_velocity_primitive.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "ai/primitive/primitive.h"
+
+class DirectVelocityPrimitive : public Primitive
+{
+   public:
+    static const std::string PRIMITIVE_NAME;
+    /**
+     * Creates a new Direct Velocity Primitive
+     *
```
[the remark this answers is not in the document]
Comment (HaomiaoZ) [reply src/thunderbots/software/ai/primitive/direct_velocity_primitive.h:18]: definitely, I modified the comment based on the suggestions, would you like to double check?

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Review by henryymliu — DISMISSED** [review]
LGTM, sorry about the late review.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
@HaomiaoZ it would be great if you could also add some unit tests for this class as well. That will give us test coverage right away, and is a good introduction to testing in our codebase.

You can use #131 as an example. Sorry for the last minute change.

**Comment by HaomiaoZ** [comment]
Ok, I will work on the unit test over the weekend, need to work on the report right now

**Comment by HaomiaoZ** [comment]
Add unit tests and get functions for individual parameters in parameters array. Also add some line of code in primitive so that the test could execute. The minor problem with the CmakeList is that all other primitive cpp files need to be included for each test (e.g. need to include move_primitive.cpp under direct velocity primitive test)

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1: Nice work!

File: src/thunderbots/software/ai/primitive/direct_velocity_primitive.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/ai/primitive/direct_velocity_primitive.h]: I might suggest moving the comment describing the functionality of the entire class to right before we declare the class (ie. on line 9), and leaving just `Creates a new Direct Velocity Primitive` for this function comment.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Only comment here is that we've dropped coverage in `primitive.cpp`, because we're not testing the new `if` case we added in the `createPrimitive` function. Adding a quick new unit test in `ai/test/primitive/primitive.cpp` should resolve this.

This slipped by in some other cases, so I've opened a PR (#139) that you can use as a reference. The relevant file is here: https://github.com/UBC-Thunderbots/Software/pull/139/files#diff-1bc8265ff2d89d3a1dc906c996260da7

**Comment by HaomiaoZ** [comment]
@garethellis0 I add the unit test in `ai/test/primitive/primitive.cpp`. A question is that are we changing function names `getParameterArray()` and `getExtraBitArray()` to `getParameters()` and `getExtraBits()`?

**Comment by MathewMacDougall** [comment]
@HaomiaoZ we shouldn't be changing the function names, they should match what is in the `primitive.h` files, which is currently `getParameterArray()` and `getExtraBitArray()`

Looks like you'll need to pull from `master` as well to resolve conflicts

**Comment by garethellis0** [comment]
@HaomiaoZ - sorry about that, don't worry about changing the function names in this PR. I plan to do that as part of the PR I linked, so I'll also change this primitive once it's merged.

**Comment by garethellis0** [comment]
Once you resolve conflicts I think the coverage should automatically update.

**Comment by HaomiaoZ** [comment]
@garethellis0 @MathewMacDougall  Evan helped me to resolve the conflict. Everything should be good now.

**Review by MathewMacDougall — APPROVED** [review]

**Comment by garethellis0** [comment]
Great work, LGTM!

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**